pkg/pillar/cmd/zedrouter/patchenvelopes.go

```diff
@@ -0,0 +1,219 @@
+		}
+		pes.Wg.Done()
+	}
+}
+
+func (pes *PatchEnvelopes) updateVolumeStatus(volumeStatus PatchEnvelopesVsCh) {
+	switch volumeStatus.Action {
+	case PatchEnvelopesVsChActionPut:
+		pes.updateExternalPatches(volumeStatus.Vs)
+	case PatchEnvelopesVsChActionDelete:
+		pes.deleteExternalPatches(volumeStatus.Vs)
+	}
+}
+
+func (pes *PatchEnvelopes) updateExternalPatches(vs types.VolumeStatus) {
+	if vs.State != types.INSTALLED {
+		return
+	}
+
+	pes.Lock()
+	defer pes.Unlock()
+
+	volumeExists := false
+	for i := range pes.completedVolumes {
+		if pes.completedVolumes[i].VolumeID == vs.VolumeID {
+			pes.completedVolumes[i] = vs
+			volumeExists = true
+			break
+		}
+	}
+	if !volumeExists {
+		pes.completedVolumes = append(pes.completedVolumes, vs)
+	}
+
+	if err := pes.processVolumeStatus(vs); err != nil {
+		pes.log.Errorf("Failed to update external patches %v", err)
+	}
+}
+
+func (pes *PatchEnvelopes) deleteExternalPatches(vs types.VolumeStatus) {
+	pes.Lock()
+	defer pes.Unlock()
+
+	i := 0
+	for _, vol := range pes.completedVolumes {
+		if vol.VolumeID == vs.VolumeID {
+			break
+		}
+		i++
+	}
+
+	pes.completedVolumes[i] = pes.completedVolumes[len(pes.completedVolumes)-1]
+	pes.completedVolumes = pes.completedVolumes[:len(pes.completedVolumes)-1]
+}
+
+func (pes *PatchEnvelopes) updateEnvelopes(peInfo []types.PatchEnvelopeInfo) {
+	pes.Lock()
+	defer pes.Unlock()
+
+	pes.envelopes = peInfo
+	if err := pes.processVolumeStatusList(pes.completedVolumes); err != nil {
+		pes.log.Errorf("Failed to update external patches %v", err)
+	}
+}
+
+func (pes *PatchEnvelopes) processVolumeStatus(vs types.VolumeStatus) error {
+	for i := range pes.envelopes {
+		for j := range pes.envelopes[i].VolumeRefs {
+			volUUID, err := uuid.FromString(pes.envelopes[i].VolumeRefs[j].ImageID)
+			if err != nil {
+				return err
+			}
+			if volUUID == vs.VolumeID {
+				blob, err := blobFromVolRef(pes.envelopes[i].VolumeRefs[j], vs)
+				if err != nil {
+					return err
+				}
+
+				if idx := types.CompletedBinaryBlobIdxByName(pes.envelopes[i].BinaryBlobs, blob.FileName); idx != -1 {
+					pes.envelopes[i].BinaryBlobs[idx] = *blob
+				} else {
+					pes.envelopes[i].BinaryBlobs = append(pes.envelopes[i].BinaryBlobs, *blob)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (pes *PatchEnvelopes) processVolumeStatusList(volumeStatuses []types.VolumeStatus) error {
+	for _, vs := range volumeStatuses {
+		if err := pes.processVolumeStatus(vs); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func blobFromVolRef(vr types.BinaryBlobVolumeRef, vs types.VolumeStatus) (*types.BinaryBlobCompleted, error) {
+	sha, err := fileutils.ComputeShaFile(vs.FileLocation)
+	if err != nil {
+		return nil, err
+	}
+	return &types.BinaryBlobCompleted{
+		FileName:     vr.FileName,
+		FileMetadata: vr.FileMetadata,
+		FileSha:      fmt.Sprintf("%x", sha),
+		URL:          vs.FileLocation,
+	}, nil
+}
+
+// since we use json in pubsub we cannot use json - tag
```

Comment:
I don't understand this comment. Is it referring to some other struct which can not be used with pubsub?

Reply:
```golang
type MyStruct struct {
   AbraCadabra string `json:"-"`
   ExposedFiled string
}

a := MyStruct{
  AbraCadabra: "My secret!",
  ExposedField: "Will be there", 
}

b := json.Marshal(a)

```

b in this case would be 
```
{
  "ExposedField" : "Will be there"
}
```

In this context we don't want to expose AllowedApps to user, we could have used json:"-" struct tag so that during marshaling it'll disappear, but we also send PatchEnvelopeInfo via pubSub from zedagent to zedrouter so adding json:"-" would make AllowedApps magically disappear from during publishing :D 

I guess it makes sense to write this in comment

Reply:
Updated the comment

Reply:
Thanks